atlasdb-workload-server/src/test/java/com/palantir/atlasdb/workload/workflow/ConcurrentTransactionRunnerTest.java

```diff
@@ -0,0 +1,132 @@
+import java.util.Optional;
+import java.util.Set;
+import java.util.concurrent.Future;
+import java.util.stream.Collectors;
+import java.util.stream.IntStream;
+import org.jmock.lib.concurrent.DeterministicScheduler;
+import org.junit.Before;
+import org.junit.Test;
+
+public class ConcurrentTransactionRunnerTest {
+    private static final int DEFAULT_TASK_MULTIPLICITY = 100;
+
+    private final IndexedTransactionTask task = mock(IndexedTransactionTask.class);
+    private final TransactionStore store = mock(TransactionStore.class);
+    private final DeterministicScheduler scheduler = new DeterministicScheduler();
+    private final ListeningExecutorService executorService = MoreExecutors.listeningDecorator(scheduler);
+    private final ConcurrentTransactionRunner runner = new ConcurrentTransactionRunner(store, executorService);
+
+    @Before
+    public void setupTask() {
+        when(task.apply(any(), anyInt()))
+                .thenAnswer(invocation ->
+                        Optional.of(createWitnessedTransactionWithStartTimestamp(invocation.getArgument(1))));
+    }
+
+    @Test
+    public void throwsIfRunningTaskNegativeNumberOfTimes() {
+        assertThatThrownBy(() -> runner.runConcurrentTransactionTask(task, -1))
+                .isInstanceOf(SafeIllegalArgumentException.class)
+                .hasMessageContaining("Tasks must be run a non-negative number of times");
+        verifyNoInteractions(task);
+    }
+
+    @Test
+    public void runsTaskAndAggregatesResults() {
+        Future<List<WitnessedTransaction>> witnessedTransactionsFuture =
+                runner.runConcurrentTransactionTask(task, DEFAULT_TASK_MULTIPLICITY);
+        assertThat(witnessedTransactionsFuture).isNotDone();
+        scheduler.runUntilIdle();
+        assertThat(witnessedTransactionsFuture).isDone();
+        List<WitnessedTransaction> witnessedTransactions = Futures.getUnchecked(witnessedTransactionsFuture);
+        assertThat(witnessedTransactions)
+                .containsExactlyElementsOf(IntStream.range(0, DEFAULT_TASK_MULTIPLICITY)
+                        .mapToObj(ConcurrentTransactionRunnerTest::createWitnessedTransactionWithStartTimestamp)
+                        .collect(Collectors.toList()));
+    }
+
+    @Test
+    public void resultFiltersOutTransactionsThatCouldNotBeWitnessed() {
+        Set<Integer> failingTasks = ImmutableSet.of(3, 7, 11, 27, 41);
+        failingTasks.forEach(
+                failingTask -> when(task.apply(any(), eq(failingTask))).thenReturn(Optional.empty()));
+
+        Future<List<WitnessedTransaction>> witnessedTransactionsFuture =
+                runner.runConcurrentTransactionTask(task, DEFAULT_TASK_MULTIPLICITY);
+        scheduler.runUntilIdle();
+        List<WitnessedTransaction> witnessedTransactions = Futures.getUnchecked(witnessedTransactionsFuture);
+        assertThat(witnessedTransactions)
+                .containsExactlyElementsOf(IntStream.range(0, DEFAULT_TASK_MULTIPLICITY)
```

Comment:
this may occasionally flake, unless `runConcurrentTransactionTask` is single threaded, so may make sense to use the `containsElementsInAnyOrderOf`?

Reply:
Good catch - yeah I tend to forget that this one does rely on the order. Will change